### PR TITLE
[Metricbeat][Kubernetes Volume] Add pvc reference to distinguish ephemeral from persistent volumes

### PR DESCRIPTION
**Pull request by herrBez**
## Proposed commit message

[Metricbeat][Kubernetes Volume] Add pvc reference to distinguish ephemeral from persistent volumes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

Create a PVC and mount the PVC as a volume in a pod, e.g. (on GKE):

```yaml
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-demo
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: standard-rwo
---
kind: Pod
apiVersion: v1
metadata:
  name: pod-demo
spec:
  volumes:
    - name: pvc-demo-vol
      persistentVolumeClaim:
       claimName: pvc-demo
  containers:
    - name: pod-demo
      image: nginx
      resources:
        limits:
          cpu: 10m
          memory: 80Mi
        requests:
          cpu: 10m
          memory: 80Mi
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
        - mountPath: "/usr/share/nginx/html"
          name: pvc-demo-vol
```

Deploy metricbeat and check that there exist documents in `kubernetes.volume` with the field `kubernetes.persistentvolumeclaim.name`.

Alternatively execute the test.

## Related issues

- Fixes https://github.com/elastic/beats/issues/6977

## Use cases

The use-case is to being able to monitor persistent volumes and ignoring ephemeral volumes for threshold alerts.

As of now there is no way to distinguish persistent from ephemeral volumes. However, the endpoint called by the module `$NODE_URL/summary/stats` does actually report a pvcReference that allows to bind the pod's volume with a corresponding pvc.

Example Response:
```json
...
 "volume": [
        {
          "time": "2024-04-09T17:34:17Z",
          "availableBytes": 31509590016,
          "capacityBytes": 31526391808,
          "usedBytes": 24576,
          "inodesFree": 1966069,
          "inodes": 1966080,
          "inodesUsed": 11,
          "name": "pvc-demo-vol",
          "pvcRef": {
            "name": "pvc-demo",
            "namespace": "default"
          }
        },
        {
          "time": "2024-04-09T17:34:17Z",
          "availableBytes": 83873792,
          "capacityBytes": 83886080,
          "usedBytes": 12288,
          "inodesFree": 502853,
          "inodes": 502862,
          "inodesUsed": 9,
          "name": "kube-api-access-h2cll"
        }
     ]
...
```

### CODE REVIEW

**Comment by herrBez** [comment]
cc @elastic/obs-ds-hosted-services

File: metricbeat/module/kubernetes/_meta/test/stats_summary.json [hunk 1]
```diff
@@ -121,7 +121,11 @@
       "inodesFree": 473551,
       "inodes": 473560,
       "inodesUsed": 9,
-      "name": "default-token-sg8x5"
+      "name": "default-token-sg8x5",
```
Comment (tetianakravchenko) [inline metricbeat/module/kubernetes/_meta/test/stats_summary.json:124]: can you please add a separate sample with the `pvcRef` to keep in the test file both examples - one for persistentvolumeclaim and one for the ephemeral volume?

**Review by tetianakravchenko — COMMENTED** [review]
could you please also add a changelog recorf?

**Comment by herrBez** [comment]
Hi,

I tried to address the raised point about the tests^^. I also tried specifying the field. What I am not sure about is that in the end I want to "reuse" `kubernetes.persistentvolumeclaim.name` which is part of the `kubernetes.state_persistentvolumeclaims` dataset. My doubt comes from the fact that we did not specify `kubernetes.pod.name`. WDYT?

**Comment by herrBez** [comment]
Hi Michael, I addressed the required changes and the build works fine

**Comment by MichaelKatsoulis** [comment]
@herrBez Can you add a screenshot showing the new field being populated to Elasticsearch? You can use a view from discovery filtering by the kubernetes.volume metricset and showing the new field.

File: metricbeat/module/kubernetes/volume/data.go [hunk 2]
```diff
@@ -75,6 +75,9 @@ func eventMapping(content []byte, logger *logp.Logger) ([]mapstr.M, error) {
 			if volume.Inodes > 0 {
 				kubernetes2.ShouldPut(volumeEvent, "fs.inodes.pct", float64(volume.InodesUsed)/float64(volume.Inodes), logger)
 			}
+			if volume.PvcRef.Name != "" {
+				kubernetes2.ShouldPut(volumeEvent, "persistentvolumeclaim.name", volume.PvcRef.Name, logger)
+			}
```
Comment (MichaelKatsoulis) [inline metricbeat/module/kubernetes/volume/data.go:80]: I have the feeling that this way the field that will be generated will be `kubernetes.volume.persistentvolumeclaim.name` . Is that what you want?

Comment (herrBez) [reply metricbeat/module/kubernetes/volume/data.go:80]: Wow, great static analysis :D.. Indeed, its creating the field you mention. I actually want to reuse the field `kubernetes.persistentvolumeclaim.name` to allow correlation, but I am honestly a little bit lost.

After some thinking I tried with:
```
kubernetes2.ShouldPut(volumeEvent, mb.ModuleDataKey+".persistentvolumeclaim.name", volume.PvcRef.Name, logger)
``` 

But it does not seem to work 🤔

Comment (herrBez) [reply metricbeat/module/kubernetes/volume/data.go:82]: Ok, the line is correct now. For some reasons I were still using the old binary.

**Comment by herrBez** [comment]
Here is the screenshot with the new version:

image

File: metricbeat/module/kubernetes/volume/data.go (hunk 2 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/kubernetes/volume/data.go:82]: I had the feeling 😂

**Review by MichaelKatsoulis — APPROVED** [review]

**Review by gizas — APPROVED** [review]

**Comment by herrBez** [comment]
Can I go ahead and merge it? Should I open a PR in the integration repository to add the field to the kubernetes.volume datastream?

**Comment by MichaelKatsoulis** [comment]
> Can I go ahead and merge it? Should I open a PR in the integration repository to add the field to the kubernetes.volume datastream?

I will merge it. About the field in the integration, not yet. We do it  after the release of beats. I will add this in our list cause we have some other fields to declare as well.

**Pull request merged**